### PR TITLE
Test statuses: do not rely on yaml_to_mux varianter plugin

### PR DESCRIPTION
**Pull request by clebergnu**
Which is an optional plugin, and the test statuses feature is at the
innermost core of Avocado.  While the original test indeed saved a lot
of duplicated code, it's code that serves a specific testing purpose,
and it's better to be simpler and more verbose and with the right (no)
dependencies.

Signed-off-by: Cleber Rosa <crosa@redhat.com>

### CODE REVIEW

**Comment by clebergnu** [comment]
[Statuses test: do not depend on optional plugins](https://trello.com/c/ZwW9lftO/1231-statuses-test-do-not-depend-on-optional-plugins)

**Comment by clebergnu** [comment]
@ldoktor no, it's not fine to rely on yaml2mux for a core test.  This is a layer violation, and this point is *not* subjective, relative or debatable.

Now, with regards to coding style used, it's just that: a coding style.  Now let me reason why I believe in the coding style that was used:
* Since this is a test for a very basic Avocado feature/foundation, it makes sense to keep it as simple as possible.  For instance, there are higher chances, with more complex implementations, that errors will be masked by unforeseen behavior of the other parts.
* To me, KISS wins over DRY here.  DRY makes a lot of sense with regards to avoiding making the same mistakes again or increasing the maintenance burden - not the case here.
* Readability.  Even though the code is longer than the previous (or suggested) implementation, it's just a 1:1 mapping that is trivial to read and maintain.
* Explicit is better than implicit (the Zen of Python is usually right ;).

Now, with that said, please re-review on the basis of *flaws* of the proposal, not on your dislike or suggestion of alternative ways of implementing it.

**Comment by clebergnu** [comment]
@ldoktor can you see that *core* is a lower level layer than *optional_plugins* ?

**Comment by ldoktor** [comment]
Yes, but functional tests depend on non-core features. So here `yaml2mux` is the external feature not provided by core (the same as isoinfo, perl, gcc, ...).

If we decided to get rid of all non-pure functional tests than yes, we'd have to get rid of this test as well. But at this point functional tests depend on many non-core programs and yaml2mux is just one of them.

**Comment by clebergnu** [comment]
On Mon, Feb 5, 2018 at 12:46 PM, Lukáš Doktor <notifications@github.com>
wrote:

> Yes, but functional tests depend on non-core features. So here yaml2mux
> is the external feature not provided by core (the same as isoinfo, perl,
> gcc, ...).
>

OK, so we are in agreement about the specific layers.  Now, when one lower
layer depends on a higher layer, that is a layer violation.  Are we in
agreement here?

It's true that some functional tests (meaning tests that live under
selftests/functional) may depend on other software.  My point, as per the
commit message is that "test statuses" is such a core Avocado
concept/principle/feature/foundation that it should be testable by itself.

Your comparison of yaml2mux with isoinfo, perl, etc is flawed because:

1) Those are not requirements to run or test avocado's core (just think
about the fact that they don't live under avocado/core)
2) Those tests are for features that usually depend on those tools
themselves
3) Those are not Avocado's own components, and as such, they do not map to
layers in the project itself.  Those relationships are horizontal, and not
vertical as in the "Avocado Core" <-> "Avocado Optional Plugins".


> If we decided to get rid of all non-pure functional tests than yes, we'd
> have to get rid of this test as well. But at this point functional tests
> depend on many non-core programs and yaml2mux is just one of them.
>

As much as possible, Avocado's core concepts/principles/features/foundations
should not depend on other anything else than Python + Stevedore + Six.  It
looks like there's a confusion here when it comes to the meaning of
"Avocado core".  I'm thinking about Avocado's core
concepts/principles/features/foundations,
which should pretty much map to "avocado/core".  It looks like you're
thinking about everything *but* optional_plugins/*.

**Comment by ldoktor** [comment]
Dne 5.2.2018 v 19:08 Cleber Rosa napsal(a):
> On Mon, Feb 5, 2018 at 12:46 PM, Lukáš Doktor <notifications@github.com>
> wrote:
> 
>> Yes, but functional tests depend on non-core features. So here yaml2mux
>> is the external feature not provided by core (the same as isoinfo, perl,
>> gcc, ...).
>>
> 
> OK, so we are in agreement about the specific layers.  Now, when one lower
> layer depends on a higher layer, that is a layer violation.  Are we in
> agreement here?
> 
> It's true that some functional tests (meaning tests that live under
> selftests/functional) may depend on other software.  My point, as per the
> commit message is that "test statuses" is such a core Avocado
> concept/principle/feature/foundation that it should be testable by itself.
> 
> Your comparison of yaml2mux with isoinfo, perl, etc is flawed because:
> 
> 1) Those are not requirements to run or test avocado's core (just think
> about the fact that they don't live under avocado/core)

/bin/echo is required to check many basic features.

> 2) Those tests are for features that usually depend on those tools
> themselves

perl parser is to check tap results. (which reminds me I should install it...)

> 3) Those are not Avocado's own components, and as such, they do not map to
> layers in the project itself.  Those relationships are horizontal, and not
> vertical as in the "Avocado Core" <-> "Avocado Optional Plugins".
> 

The point is we are not interested in testing the optional plugin, we use it as a tool to get what we need in a nice, systematic, easy to maintain way. In this matter it's an external black box to get what we need. Would we break layers by depending on aexpect? Or Avocado-vt?

In unittest yes, it'd be a violation as unittests focus on testing individual components (and even here many projects are breaking layers because developing and maintaining in-test alternatives are more costly), but we are talking about functional tests. These don't focus on individual parts but on the overall functionality.

> 
>> If we decided to get rid of all non-pure functional tests than yes, we'd
>> have to get rid of this test as well. But at this point functional tests
>> depend on many non-core programs and yaml2mux is just one of them.
>>
> 
> As much as possible, Avocado's core concepts/principles/features/foundations
> should not depend on other anything else than Python + Stevedore + Six.  It
> looks like there's a confusion here when it comes to the meaning of
> "Avocado core".  I'm thinking about Avocado's core
> concepts/principles/features/foundations,
> which should pretty much map to "avocado/core".  It looks like you're
> thinking about everything *but* optional_plugins/*.
> 

In unittests yes, in complex functional tests I don't have a problem of external or optional component being required if that improves the test maintainability. Besides to keep the coverage we'd have to provide similar test to the yaml2mux plugin.

Anyway I feel like we're moving in circles so I think we need some 3rd person opinion... @apahim @lmr what are your opinions?

**Comment by clebergnu** [comment]
On Tue, Feb 6, 2018 at 4:41 AM, Lukáš Doktor <notifications@github.com>
wrote:

> Dne 5.2.2018 v 19:08 Cleber Rosa napsal(a):
> > On Mon, Feb 5, 2018 at 12:46 PM, Lukáš Doktor <notifications@github.com>
> > wrote:
> >
> >> Yes, but functional tests depend on non-core features. So here yaml2mux
> >> is the external feature not provided by core (the same as isoinfo, perl,
> >> gcc, ...).
> >>
> >
> > OK, so we are in agreement about the specific layers. Now, when one lower
> > layer depends on a higher layer, that is a layer violation. Are we in
> > agreement here?
> >
> > It's true that some functional tests (meaning tests that live under
> > selftests/functional) may depend on other software. My point, as per the
> > commit message is that "test statuses" is such a core Avocado
> > concept/principle/feature/foundation that it should be testable by
> itself.
> >
> > Your comparison of yaml2mux with isoinfo, perl, etc is flawed because:
> >
> > 1) Those are not requirements to run or test avocado's core (just think
> > about the fact that they don't live under avocado/core)
>
> /bin/echo is required to check many basic features.
>
>
And I stick to the point that we *shouldn't* depend on anything but Python
+ Stevedore + Six for the *core* features of Avocado.  Please pause, and
think about those "many basic features" you mentioned.  Should those work,
say, on Windows with Python + Stevedore + Six? O should they only be tested
if `/bin/echo` is available?  In fact, we should really have substitutes
for `/bin/echo` (Python scripts?) to improve the coverage and portability.

In summary, the fact that `/bin/echo` is used to test many basic features
is not a justification to keep doing so.


> > 2) Those tests are for features that usually depend on those tools
> > themselves
>
> perl parser is to check tap results. (which reminds me I should install
> it...)
>
>
Which again, doesn't fall, *at all*, in what I described as "Avocado's core
concepts/principles/features/foundations, which should pretty much map to
"avocado/core".

> 3) Those are not Avocado's own components, and as such, they do not map to
> > layers in the project itself. Those relationships are horizontal, and not
> > vertical as in the "Avocado Core" <-> "Avocado Optional Plugins".
> >
>
> The point is we are not interested in testing the optional plugin, we use
> it as a tool to get what we need in a nice, systematic, easy to maintain
> way. In this matter it's an external black box to get what we need. Would
> we break layers by depending on aexpect? Or Avocado-vt?
>
>
You're right, in this specific example, we're not testing the optional
plugin.  We're testing, via a *functional test*,
concepts/principles/features/foundations such as the one ones implemented
in avocado/core/status.py, avocado/core/runner.py.

The point is that, if we can get the same coverage without resorting to any
other external dependencies, we have a broader and more stable coverage.

The situation we have here, is something like if we were building an
assembler, and relying on a compiler for a test because we don't want to
write machine code.

Please, pretty please, think about this proposal in terms of what is
undeniably broken with it.  I've already listed some of reasons for this
work, so if you could, please, list why it's *broken*, then we can close it.


> In unittest yes, it'd be a violation as unittests focus on testing
> individual components (and even here many projects are breaking layers
> because developing and maintaining in-test alternatives are more costly),
> but we are talking about functional tests. These don't focus on individual
> parts but on the overall functionality.
>
> >
> >> If we decided to get rid of all non-pure functional tests than yes, we'd
> >> have to get rid of this test as well. But at this point functional tests
> >> depend on many non-core programs and yaml2mux is just one of them.
> >>
> >
> > As much as possible, Avocado's core concepts/principles/features/
> foundations
> > should not depend on other anything else than Python + Stevedore + Six.
> It
> > looks like there's a confusion here when it comes to the meaning of
> > "Avocado core". I'm thinking about Avocado's core
> > concepts/principles/features/foundations,
> > which should pretty much map to "avocado/core". It looks like you're
> > thinking about everything *but* optional_plugins/*.
> >
>
> In unittests yes, in complex functional tests I don't have a problem of
> external or optional component being required if that improves the test
> maintainability. Besides to keep the coverage we'd have to provide similar
> test to the yaml2mux plugin.
>

You don't have a problem, but the problem exists.  And again, this work is
fixing existing issues.  You *not* having a problem with it, is not a
reason to not accept it.  And, if you think this coverage is *also* needed
in yaml2mux, we should move/copy this implementation to that plugin instead.


>
> Anyway I feel like we're moving in circles so I think we need some 3rd
> person opinion... @apahim @lmr what are your opinions?
>
>
I'd love to hear the opinion of others, but I guess one point cannot be
ignored: *you* have to tell why this broken.

**Comment by ldoktor** [comment]
The reason I nacked this patch is the decrease of readability of this test which means harder maintainability of this test. In case of a simple change it'd require many modifications and typos would be easy to introduce. Also previously you could easily see the connection between variant, it's execution and the expected results but now you have to go through many redundant classes with really minimal differences in them which is again prone to typos or just manual fixing instead of seeing what behavior changed.

I don't say it's makes maintenance impossible, but way harder, especially the part about connecting expected results to what will be executed is quite big deal to me. We've had selftest fixes that were actually breaking the expected results (stdout recorder) because the test was not that readable.

Anyway the main question I see here is, whether we do want to have external deps in "functional" tests or not. I don't have a problem with them, because functional tests are complex cases that are expected to utilize the whole workflows. Note that in "unit" tests we should avoid such deps as much as possible.

As for the current status, this is the results without yaml2mux plugin:

```
....................................................................................................................................sssss.......................................................................................................F.E................EEEEEEEEEEEEEE..............EE..........................EEF...EE.FF.F..F.E.........FF.FFF..................................E............................................FFFFFFFFFFF..................................
...
----------------------------------------------------------------------
Ran 472 tests in 248.834s

FAILED (failures=22, errors=23, skipped=5)
```

and I really don't think this dependency is a problem. If you want to distinguish pure and complex tests, we can introduce `selftests/functional_with_optional_plugins/` location for integration tests with optional plugins, but for me `functional` already implies complex cases with various deps.

**Comment by clebergnu** [comment]
@ldoktor so you stand that the only reason for nacking is that "this patch is the decrease of readability of this test which means harder maintainability of this test.", right?

**Comment by ldoktor** [comment]
For this PR yes and for the big discussion that raised about using optional plugins (or other non-basic deps) in "functional" tests I stick to my opinion.

That means you can introduce better version that will perhaps address the readability and unblock this NACK, but before spending even more time on fixing all the 45 tests we should clearly sort out the question of whether it's acceptable, wanted or prohibited to use those deps in functional tests.

**Comment by clebergnu** [comment]
@ldoktor funny, because previously you've said, and I quote: 

> As for the style I wrote there "consider", not that it's the main issue.

To me, you're now holding on to the style aspect[1] because it's the only thing you can now hold on to.  And blocking something based on what you initially asked to "consider" changing, is pretty bad for the health of this or any other collaborative project.

**Comment by apahim** [comment]
The reviewer job is to show the issues with the PR. On that:
- the readability is good enough
- the PR does not decrease the coverage
- the dependency removal does not affect the overal test result
Personally, I don't think external dependencies are to be forbidden in functional tests, but I see the value in a best effort mindset in that direction. With that said, if this PR does not introduce a requirement for non external dependencies in functional tests, which I don't think it does, I'd accept it.
Now that you have my opinion, please finish this discussion: @cleber, if the nack stands, just close the PR and move on. @ldoktor, if my opinion makes you to change your mind, please finish the code review, merge it and move on. Move on guys.

**Comment by ldoktor** [comment]
And it still applies because even with the style change I imagine it'd be still too complex (maybe not, I haven't seen it) so the main issue with maintainability would be still there, which was the first (and main) point. Anyway if you improve the code to the fact it clearly maps the code to expected results as is now I'll ack it. Still I'd like to get some consensus about functional selftests as it's not only about this single case.

**Comment by clebergnu** [comment]
@ldoktor it has been hard to follow your line of thought.  You've said before that you didn't believe there was a layer violation, and that was why you were nacking it:

> So to explain a bit better the previous comment the nack was to the fact that it's level violation...

Please answer, as much as you can with a straight answer, to make things clearer:

1) do you now see the layer violation?
2) do you understand the purpose of this proposal?
3) do you stick to your nack on the basis of what you consider "maintanability" (and I consider style)?

You last comment makes me believe that, if I rewrite the code, and you still think it's "still too complex", you'll hold your NACK.  So, I'd like to get these answers first before putting more time into this.

FIY, to give you more background, this started with me testing, functionally, Avocado on Python 3.  yaml_to_mux has issues under Python 3, and because of that, Avocado's *core* tests were failing.  To me, that doesn't make sense and translates clearly to a layer violation.  That's a real issue, and you're blocking *a* fix (not the only possible one, of course), on what it looks to me like unstable (as in changes) and weak (style? "maintainability"?) points.

@apahim I'm sorry but I can't let go on this.  As I explained before, it's a issue that's blocking the Python 3 port.

**Comment by apahim** [comment]
Ok, so if the python3 support is depending on this and there are *no* drawbacks in this change other than the alleged readability depreciation, I don't understand why we are discussing this. Such a waste of time.

**Comment by ldoktor** [comment]
I was not aware this is because py3 is not able to ran this test. The question is do you see a solution for the yaml2mux? I saw some fixes recently. If you think we can fix yaml in reasonable timeframe I'd just forget about this because I nor Amador see a problem with this external dependency. Simple skipIf would do.

If you don't see us fixing yaml2mux any time soon then we need to assess the benefits in having this test py3 compatible. Yes, it'd be nice but given we are running it on py2, it does not brings any new code coverage (see below) and the fix decreases the maintainability I'd stick to the old way + skip. If the code readability improves so it overcomes the benefits we can merge it but otherwise it's not worth it. (I mean we'll have to keep fixing this in case of change, right now we can show it users and say those are expected behaviors but the proposed version does not do that).

To wrap it up let's just (in general) keep the optional+external deps for functional tests. As for this PR I'd consider merging it only if the benefits overcome the maintainability issue, therefor either the yaml2mux is not fixable any time soon and you believe this test is very important. On the other side of the equation we have the new code readability so improved code readability might lower the need for this test to be that important but if the yaml2mux is to be fixed soon I'd just close this PR and focus on that.

----------------------------------------------------

I looked at code coverage and the `test_statuses` does not adds any new coverage to `test_basic` itself:

```
$ coverage erase
$ AVOCADO_CHECK_LEVEL=2 UNITTEST_AVOCADO_CMD="coverage run -p --include 'avocado/*' $ ./scripts/avocado" coverage run -p --include "avocado/*" ./selftests/functional/test_basic.py
$ coverage combine .coverage*
$ coverage report -m --include "avocado/core/*"
TOTAL                                4017   1557    61%
$ AVOCADO_CHECK_LEVEL=2 UNITTEST_AVOCADO_CMD="coverage run -p --include 'avocado/*' ./scripts/avocado" coverage run -p --include "avocado/*" ./selftests/functional/test_statuses.py
$ coverage combine .coverage*
$ coverage report -m --include "avocado/core/*"
TOTAL                                4017   1557    61%
```
Which obviously doesn't mean the test doesn't check anything new but it means it does not executes different code-paths and we do have some basic test statuses (PASS, FAIL, INTERRUPTED, ...) checked elsewhere.

**Comment by clebergnu** [comment]
@ldoktor would you *please*, address the questions I made?

**Comment by ldoktor** [comment]
@apahim I don't see this as a waste of time, there is a very important question about what can go to functional tests. Once this question is addressed we can use it as precedence which avoids such issues in the future.

Development is not black&white, it's full of gray areas and this one is one of them. We have a very nice, almost documentation-like selftest, on the other hand we have a problem with py3 and a conceptual question of what-can-go-to-functional-tests. So now it's about evaluating the pros and cons.

**Comment by ldoktor** [comment]
Dne 7.2.2018 v 12:31 Cleber Rosa napsal(a):
> @ldoktor it has been hard to follow your line of thought.  You've said before that you didn't believe there was a layer violation, and that was why you were nacking it:
> 
>> So to explain a bit better the previous comment the nack was to the fact that it's level violation...
> 
> Please answer, as much as you can with a straight answer, to make things clearer:
> 
> 1) do you now see the layer violation?

no, in my understanding functional tests can depend on anything to make it possible to write them or to avoid unnecessary complex structures.

> 2) do you understand the purpose of this proposal?

First I thought it's because of layer violation (which I don't see), now I think it's because of py3 test coverage which I think should be addressed by the dependency if possible, skipped if not that important or using this if there is no other choice.

> 3) do you stick to your nack on the basis of what you consider "maintanability" (and I consider style)?

Yes, it's not just coding style, it's the fact that now it clearly demonstrate the actions and expected behavior. This version doesn't do that and any change to expected behavior brings many changes that can lead to broken selftests (and we've already had that).

> 
> You last comment makes me believe that, if I rewrite the code, and you still think it's "still too complex", you'll hold your NACK.  So, I'd like to get these answers first before putting more time into this.

It's an equation, if the code can somehow show the expected Avocado behavior then I'll withdraw my NACK. I can also ACK it if there is no other way in fixing yaml2mux on py3 and you think it's necessary to run this test on py3 as that changes the scales.

> 
> FIY, to give you more background, this started with me testing, functionally, Avocado on Python 3.  yaml_to_mux has issues under Python 3, and because of that, Avocado's *core* tests were failing.  To me, that doesn't make sense and translates clearly to a layer violation.  That's a real issue, and you're blocking *a* fix (not the only possible one, of course), on what it looks to me like unstable (as in changes) and weak (style? "maintainability"?) points.
> 
> @apahim I'm sorry but I can't let go on this.  As I explained before, it's a issue that's blocking the Python 3 port.
>

**Comment by clebergnu** [comment]
On Wed, Feb 7, 2018 at 8:06 AM, Lukáš Doktor <notifications@github.com>
wrote:

> Dne 7.2.2018 v 12:31 Cleber Rosa napsal(a):
> > @ldoktor it has been hard to follow your line of thought. You've said
> before that you didn't believe there was a layer violation, and that was
> why you were nacking it:
> >
> >> So to explain a bit better the previous comment the nack was to the
> fact that it's level violation...
> >
> > Please answer, as much as you can with a straight answer, to make things
> clearer:
> >
> > 1) do you now see the layer violation?
>
> no, in my understanding functional tests can depend on anything to make it
> possible to write them or to avoid unnecessary complex structures.
>
>
My question was not if you think it's OK to depend on anything (including
crossing layers) but whether you see that a test that tests
`avocado/core/*` code depending on `optional_plugins/*` crosses layers.


> > 2) do you understand the purpose of this proposal?
>
> First I thought it's because of layer violation (which I don't see), now I
> think it's because of py3 test coverage which I think should be addressed
> by the dependency if possible, skipped if not that important or using this
> if there is no other choice.
>
>
Oh, you've answered it here: you don't see the layer violation.  And you
want to see this done differently.


> > 3) do you stick to your nack on the basis of what you consider
> "maintanability" (and I consider style)?
>
> Yes, it's not just coding style, it's the fact that now it clearly
> demonstrate the actions and expected behavior. This version doesn't do that
> and any change to expected behavior brings many changes that can lead to
> broken selftests (and we've already had that).
>
>
Can you elaborate on this? The `EXPECTED_RESULTS`, where I believe is where
big part of the "expeted behavior" is in the code, was kept the same.  The
general approach of the test wasn't changed.  Basically this takes out the
automatic (and more implicit) generation of tests and introduces a manual
set of tests (more explicit) while dropping the optional_plugin usage.

Now, if a change of behavior occurs, it *should* lead to broken selftests,
so I don't follow when you say: "any change to expected behavior brings
many changes that can lead to broken selftests", unless you're saying that
one *may*, in the future, add bugs to this or other test.  And I find it
amazing that we're debating what *may happen* if developers add bugs to
test code...



> >
> > You last comment makes me believe that, if I rewrite the code, and you
> still think it's "still too complex", you'll hold your NACK. So, I'd like
> to get these answers first before putting more time into this.
>
> It's an equation, if the code can somehow show the expected Avocado
> behavior then I'll withdraw my NACK. I can also ACK it if there is no other
> way in fixing yaml2mux on py3 and you think it's necessary to run this test
> on py3 as that changes the scales.
>
>
What do you mean by "somehow show the expected Avocado behavior" ??? I'm
trully puzzled here.  This is the code, mostly unchanged from the previous
version:

EXPECTED_RESULTS = {'SkipSetup.test': ('SKIP',
                                       ['setup pre',
                                        "[WARNING: self.skip() will be "
                                        "deprecated. Use 'self.cancel()' "
                                        "or the skip decorators]"]),
...

Which, again, is the same approach used before, the only change is that
intead of a variant name, we have a test name (with the same level of
descriptiveness).  This is the diff from the current version:

-EXPECTED_RESULTS = {'skip-setup-d304': ('SKIP',

So please enlighten me here.

**Comment by ldoktor** [comment]
I'm sorry for not answering the question correctly. So in general yes, `avocado` must never depend on `optional_plugins` and we keep improving on that. The question whether `selftests` can is a different one and here I'd go with `no` for `unit` (unless there is no feasible way around that) and `yes` for `functional`.

By the clear mapping I mean:

```python
from avocado import Test


class StatusTest(Test):

    def setUp(self):
        self.log.info('setup pre')
        if self.params.get('location') == 'setUp':
            exec(self.params.get('command'))
        self.log.info('setup post')

    def test(self):
        self.log.info('test pre')
        if self.params.get('location') == 'test':
            exec(self.params.get('command'))
        self.log.info('test post')

    def tearDown(self):
        self.log.info('teardown pre')
        if self.params.get('location') == 'tearDown':
            exec(self.params.get('command'))
        self.log.info('teardown post')
```
which on first sight gives me a notion of what can happen and

```yaml
locations: !mux
    setup:
        location: setUp
    test:
        location: test
    teardown:
        location: tearDown
commands: !mux
    skip:
        command: self.skip()
    cancel:
        command: self.cancel()
    fail:
        command: self.fail()
    warn:
        command: self.log.warn('')
    exception:
        command: raise Exception
    exit:
        command: import sys; sys.exit(-1)
    kill:
        !filter-only : /run/locations/test
        command: import os; os.kill(os.getpid(), 9)
```
which tells me what all events can happen and where do we care about their results. Then I can easily see what is the expected behavior but with nearly 20 classes I don't get this overview.

Now adding a different action or changing the place to execute the action currently means a simple change but with so many classes it means producing yet another set of classes, which is prone to forgetting about some or fixing it incorrectly (based on broken implementation). That's what is the main issue for me and it can be minimized, but without dynamic code you can't easily get this coverage without multiple classes. That's why I said if the benefits overcome the decrease of readability.

Would you please also answer my question about yaml2mux on py3? Will it be possible any time soon?

**Comment by clebergnu** [comment]
@ldoktor your explanation to me makes it clear that it's really a matter of how different brains work.  I find this:

```python
class SkipSetup(Test):

    def setUp(self):
        self.log.info('setup pre')
        self.skip()
        self.log.info('setup post')

    def test(self):
        self.log.info('test pre')
        self.log.info('test post')

    def tearDown(self):
        self.log.info('teardown pre')
        self.log.info('teardown post')
```

Easier to grasp than:

```python
class StatusTest(Test):

    def setUp(self):
        self.log.info('setup pre')
        if self.params.get('location') == 'setUp':
            exec(self.params.get('command'))
        self.log.info('setup post')

    def test(self):
        self.log.info('test pre')
        if self.params.get('location') == 'test':
            exec(self.params.get('command'))
        self.log.info('test post')

    def tearDown(self):
        self.log.info('teardown pre')
        if self.params.get('location') == 'tearDown':
            exec(self.params.get('command'))
        self.log.info('teardown post')
```

Because:
* It's more code to read (at once)
* I have to refer to a secondary file (the varianter YAML)
* Secondary file is in a different format
* Secondary *YAML* file contains *Python* code

I'm not saying my brain works *right* and yours works *wrong*, just that we have different *preferences*.

About your question about fixing yaml_to_mux on Python 3: it will have to be done, and I do sincerely plan to finish this port ASAP.  I can't predict how long it will take until I dig deeper into it.  Again, you're proposing a **different** way to approach this work, and I have to say that I find this pretty inconsiderate with regards to the energy that was already put here.

**Comment by ldoktor** [comment]
@clebergnu, I can withdraw my NACK if the reason is py3 selftest coverage (which means fixing the commit message) but I'm not sure what happens next. Do you intend to modify all tests that use yaml2mux? If so than we'd have to create those tests in `optional_plugins` to keep the coverage (obviously not all of them, but you need some coverage), which means executing similar things not just multiple, but many times only to say in a month that we have py3 compatibility and the old tests would work as well. So if it makes you happy I can merge this (with a different commit message as we don't mind the optional_plugin deps) but I'd way much prefer fixing the yaml2mux on py3 and until then `skipIf`.

**Comment by clebergnu** [comment]
@ldoktor my approach towards the Python 3 port (currently core) led me into this, but **should**  Python 3 port be mentioned in every change the happens during the process? I don't think so.  I think this change stands on its own merit.

Still, I'm all in favor of good commits messages, so let's record this for posterity.  I'm updating the commit message.

With regards to your questions on further eliminating layer violations, I mentioned that as much as possible it should be done indeed.

**Comment by clebergnu** [comment]
Closing in favor of #2435
